tests/integration/test_grub.py

```diff
@@ -0,0 +1,201 @@
+    """Test that charm can do multiple updates and update it's own configuration."""
+    # charms using this config to make update
+    configs = [
+        {"GRUB_TIMEOUT": "0"},
+        {
+            "GRUB_TIMEOUT": "0",
+            "GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G",
+        },
+        {"GRUB_TIMEOUT": "1"},
+    ]
+    # charms configs in time
+    exp_charms_configs = [
+        {"GRUB_TIMEOUT": "0"},
+        {
+            "GRUB_TIMEOUT": "0",
+            "GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G",
+        },
+        {
+            "GRUB_TIMEOUT": "1",
+            "GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G",
+        },
+    ]
+    exp_main_config = {
+        "GRUB_TIMEOUT": "1",
+        "GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G",
+    }
+    grub_conf = grub.Config("test-charm")
+
+    for config, exp_conf in zip(configs, exp_charms_configs):
+        grub_conf.update(config)
+        assert exp_conf == grub_conf
+        assert exp_conf == grub._load_config(grub_conf.path)
+
+    # check the main config
+    assert exp_main_config == grub._load_config(grub.GRUB_CONFIG)
+
+
+@pytest.mark.parametrize(
+    "config_1, config_2",
+    [
+        ({"GRUB_TIMEOUT": "0"}, {"GRUB_TIMEOUT": "0"}),
+        (
+            {"GRUB_TIMEOUT": "0"},
+            {"GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G"},
+        ),
+        (
+            {"GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G"},
+            {"GRUB_TIMEOUT": "0"},
+        ),
+    ],
+)
+def test_two_charms_no_conflict(config_1, config_2):
+    """Test two charms update GRUB configuration without any conflict."""
+    for name, config in [("test-charm-1", config_1), ("test-charm-2", config_2)]:
+        grub_conf = grub.Config(name)
+        grub_conf.update(config)
+        assert config == grub._load_config(grub_conf.path)
+
+    # check the main config
+    assert {**config_1, **config_2} == grub._load_config(grub.GRUB_CONFIG)
+
+
+@pytest.mark.parametrize(
+    "config_1, config_2",
+    [
+        ({"GRUB_TIMEOUT": "0"}, {"GRUB_TIMEOUT": "1"}),
+        (
+            {"GRUB_TIMEOUT": "0"},
+            {
+                "GRUB_TIMEOUT": "1",
+                "GRUB_CMDLINE_LINUX_DEFAULT": "$GRUB_CMDLINE_LINUX_DEFAULT hugepagesz=1G",
+            },
+        ),
+    ],
+)
+def test_two_charms_with_conflict(config_1, config_2):
+    """Test two charms update GRUB configuration with conflict."""
+    # configure charm 1
+    grub_conf_1 = grub.Config("test-charm-1")
+    grub_conf_1.update(config_1)
+    assert config_1 == grub._load_config(grub_conf_1.path)
+
+    # configure charm 2
+    grub_conf_2 = grub.Config("test-charm-2")
+    with pytest.raises(grub.ValidationError):
+        grub_conf_2.update(config_2)
+
+    assert grub_conf_2.path.exists() is False
```

Comment:
`assert not grub_conf_2.path.exists()` would be the Pythonic way to spell this. (Similar elsewhere.)